### PR TITLE
LID-2: Model, loss and pooling modules

### PR DESCRIPTION
**Pull request by Qingzheng-Wang**
## What did you change?

This PR introduces the core model components and loss functions for language identification (LID), located under the `espnet2/lid` directory.

Specifically, it includes:

- **Loss functions** (`espnet2/lid/loss`):
  - Adapted from ESPnet-SPK losses.
  - Modified to support LID-specific training by adding language classification accuracy computation and returning predicted language IDs.

- **Pooling functions** (`espnet2/lid/pooling`):
  - Adapted from ESPnet-SPK pooling layers.
  - Modified to ensure that mean, statistical, and attention-based pooling operate only on valid (unpadded) feature lengths.

- **`espnet2/lid/espnet_model`**

---

## Why did you make this change?

These modules are adapted from ESPnet-SPK components and revised to support language identification, including changes in output structure and padding handling.

---

## Is your PR small enough?

Yes

---

## Additional Context

- Part 2 of the LID system.
- Depends on: #6155

### CODE REVIEW

**Comment by sw005320** [comment]
This pull request introduces a new language identification (LID) model to the ESPnet framework, along with supporting modules for feature extraction, pooling, and loss computation. The changes primarily focus on adding the `ESPnetLIDModel` class, implementing various loss functions, and providing pooling mechanisms for aggregating frame-level features into utterance-level embeddings.

### Language Identification Model:
* [`espnet2/lid/espnet_model.py`](diffhunk://#diff-0b495cb7ee0dbdaac7585af77d32a48b962a0052aee2489314f290b3c7672b01R1-R158): Added the `ESPnetLIDModel` class, which supports language identification and embedding extraction. It integrates frontend, encoder, pooling, and loss modules for processing speech data.

### Loss Functions:
* [`espnet2/lid/loss/aamsoftmax.py`](diffhunk://#diff-16964c8a9ed1077ad359e43bcbc018daf547f6f1a372a57936f475004bdb9a45R1-R112): Implemented the AAMSoftmax loss, which uses additive angular margin for better feature discrimination.
* [`espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py`](diffhunk://#diff-f70fa8c50f079bc42c8f184f8e4ac8920e7cab9c58a7fb92897480bd5844f8d6R1-R191): Added a loss function combining ArcFace, sub-center enhancement, and inter-topK penalty for improved intra-class variation handling and hard negative sample penalization.
* [`espnet2/lid/loss/softmax.py`](diffhunk://#diff-177284531320a2a6b9b84cf605f3dae2d87b15e863cebed466f2a555b3107992R1-R60): Introduced a standard Softmax loss for language identification tasks.
* [`espnet2/lid/loss/abs_loss.py`](diffhunk://#diff-090487d95db2be05374956f1b4f27b523433056f5a3b326784b7902dee86b8a5R1-R16): Created an abstract base class `AbsLoss` to standardize loss function interfaces.

### Pooling Mechanisms:
* [`espnet2/lid/pooling/chn_attn_stat_pooling.py`](diffhunk://#diff-56f76c54d8563c1bcc682a3ccebc86795c6ea7ec3e17abc8929bbf85793a3593R1-R97): Added channel-attentive statistical pooling, which aggregates frame-level features into utterance-level embeddings using attention mechanisms.
* [`espnet2/lid/pooling/mean_pooling.py`](diffhunk://#diff-0f2f120a6fbafc838009ca9e9f591c3fd0f4503a670ac53ce236e84e15c009ffR1-R44): Implemented mean pooling for averaging frame-level features into a single utterance-level feature.
* [`espnet2/lid/pooling/abs_pooling.py`](diffhunk://#diff-9ee105fda5620769748db825d72e12e30361abc752e18148c4e935a585fc85b0R1-R14): Created an abstract base class `AbsPooling` to define pooling interfaces.

**Comment by sw005320** [comment]
Please fix the conflict.

File: espnet2/lid/espnet_model.py
Comment (sw005320) [inline espnet2/lid/espnet_model.py:1]: It's more like a note or comment.
How do we handle the code-switching case?
Can we extend utt2lang to have multiple languages?

File: espnet2/lid/loss/aamsoftmax.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+from typing import Optional, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from espnet2.lid.loss.abs_loss import AbsLoss
+
+
+class AAMSoftmax(AbsLoss):
```
Comment (sw005320) [inline espnet2/lid/loss/aamsoftmax.py]: This sounds duplicated with `espnet2/spk/loss/aamsoftmax.py`?
Why do we need to have a new function?
Please explain it and add a comment if needed.

File: espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py
Comment (sw005320) [inline espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py:1]: ditto
why do we need to duplicate it?
please explain/clarify how it differs to avoid some confusions.

File: espnet2/lid/espnet_model.py
Comment (Qingzheng-Wang) [reply espnet2/lid/espnet_model.py:1]: Currently it only supports one utterance corresponding to one language label.

File: espnet2/lid/loss/softmax.py
Comment (sw005320) [inline espnet2/lid/loss/softmax.py:1]: ditto

File: espnet2/lid/pooling/abs_pooling.py
Comment (sw005320) [inline espnet2/lid/pooling/abs_pooling.py:1]: ditto
Please explain why we need additional pooling functions

File: espnet2/lid/pooling/chn_attn_stat_pooling.py
Comment (sw005320) [inline espnet2/lid/pooling/chn_attn_stat_pooling.py:1]: ditto
Please explain why we need additional pooling functions

File: espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py
Comment (Qingzheng-Wang) [reply espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py:1]: The main differences are:

1. This implementation returns predicted labels (via argmax) during inference, enabling LID predictions.
2. It also computes and returns accuracy during training and evaluation, making it easier to monitor model performance.

**Comment by sw005320** [comment]
@Jungjee, can you review this PR?
I'm actually fine with having some duplicated functions, but I want to make sure.

File: espnet2/lid/pooling/chn_attn_stat_pooling.py
Comment (Qingzheng-Wang) [reply espnet2/lid/pooling/chn_attn_stat_pooling.py:1]: Compared to the SPK version (`espnet2/spk/pooling/*`), these implementations explicitly handle feature lengths, ensuring that mean and standard deviation pooling are computed only over valid (unpadded) frames.

Comment (Qingzheng-Wang) [reply espnet2/lid/pooling/chn_attn_stat_pooling.py:1]: I have added comments explaining this.

File: espnet2/lid/pooling/stat_pooling.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class StatsPooling(AbsPooling):
+    """Aggregates frame-level features to single utterance-level feature.
+
+    Note:
+        Compared to the SPK version (`espnet2/spk/pooling/stat_pooling.py`),
+        this implementation explicitly handles feature lengths,
+        ensuring that mean and standard deviation pooling are
+        computed only over valid (unpadded) frames.
```
Comment (sw005320) [inline espnet2/lid/pooling/stat_pooling.py]: Do you mean that the origional SPK's pooling does not care about the padded region?

@Jungjee, if so, I think it's better to use this for SPK.
Can you check it?

Comment (Qingzheng-Wang) [reply espnet2/lid/pooling/stat_pooling.py]: Yes, the original SPK pooling aggregated over all frames, including padding frames.

**Comment by sw005320** [comment]
@Jungjee, can you confirm it?
If the padding issue is true, this would be an improvement point of your work.

**Comment by sw005320** [comment]
@holvan, can you check if these fixed pooling functions are working in your SPK experiments?
If it has some improvements (or it does not degrade the performance), we should replace pooling functions in both SPK and LID with these modified functions.

**Comment by sw005320** [comment]
> @Jungjee, can you confirm it? If the padding issue is true, this would be an improvement point of your work.

@Jungjee, this is a reminder.
Can you take a look at this?
This is important for your SID implementation, and we need to review it carefully and make the necessary corrections.
I don't want to have twisted situations (e.g.,  LID has the correct implementation while SID has some bugs in pooling).

File: espnet2/lid/loss/aamsoftmax.py (hunk 1 above)
Comment (Jungjee) [reply espnet2/lid/loss/aamsoftmax.py]: +1.
I don't see critical difference compared to `espnet2/spk/loss/aamsoftmax.py`. 
Is there a reason why you're introducing this file instead of importing `espnet2/spk/loss/aamsoftmax.py`?

File: espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
+# Code adapted from:
```
Comment (Jungjee) [inline espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py]: ditto. Please let us know what is different to the original function in spk side and whether you can enhance that code instead of creating a nearly-duplicate.

File: espnet2/lid/loss/abs_loss.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from abc import abstractmethod
```
Comment (Jungjee) [inline espnet2/lid/loss/abs_loss.py]: ditto. I am not currently sold on adding `espnet2/lid/loss` in general.

File: espnet2/lid/pooling/abs_pooling.py
Comment (Jungjee) [reply espnet2/lid/pooling/abs_pooling.py:1]: +1

**Comment by sw005320** [comment]
Thanks @Jungjee!
@Qingzheng-Wang, can you reflect the comments?

File: espnet2/lid/loss/aamsoftmax.py (hunk 1 above)
Comment (Qingzheng-Wang) [reply espnet2/lid/loss/aamsoftmax.py]: The main differences are:

1. This implementation returns predicted labels (via argmax) during inference, enabling output with LID predictions during inference.
2. It also computes and returns accuracy during training and evaluation, making it easier to monitor model performance.

Although the core principle of AAMSoftmax remains the same as in spk, this version uses a different output interface specific for the LID task. This modification is necessary for my design of the LID inference pipeline (#6157 ), which outputs the LID predictions.

File: espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py (hunk 3 above)
Comment (Qingzheng-Wang) [reply espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py]: My explanation is:

I want to return the LID predictions during inference (the LID inference pipeline is #6157 ). The predictions are obtained by applying argmax to the cosine similarity scores:

```
cosine = F.linear(
    F.normalize(input), F.normalize(self.weight)
)  # Output: (batch_size, nclasses)

pred_lids = torch.argmax(cosine, dim=1)
```

So, I rewrite the forward functions of original losses.

File: espnet2/lid/loss/aamsoftmax.py (hunk 1 above)
Comment (sw005320) [reply espnet2/lid/loss/aamsoftmax.py]: OK, please embed such a comment and avoid future confusions by the users

Comment (Jungjee) [reply espnet2/lid/loss/aamsoftmax.py]: @Qingzheng-Wang @sw005320 , to me if it's merely additional returns, I still believe we could avoid adding dupliated functions but instead add these chages to ESPnet-SPK module side and update ESPnet-SPK to accomodate this change.

File: espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py (hunk 3 above)
Comment (Jungjee) [reply espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py]: I'm not sure if this difference is significant enough so that we need a new function/class..

File: espnet2/lid/loss/aamsoftmax.py (hunk 1 above)
Comment (Qingzheng-Wang) [reply espnet2/lid/loss/aamsoftmax.py]: But please consider, LID is designed only for classification, but SPK is designed for speaker embedding extraction.

File: espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py (hunk 3 above)
Comment (Qingzheng-Wang) [reply espnet2/lid/loss/aamsoftmax_subcenter_intertopk.py:1]: I sincerely apologize, but I have already explained my considerations in detail.
It is possible to directly modify the SPK task, but is that really reasonable? From my understanding, the LID task primarily focuses on classification, while SPK provides embeddings that can support many more tasks, such as verification, in addition to classification.

Why would adding a separate loss implementation be considered a simple duplication? My intention was simply to ensure that the LID task is implemented correctly and does not interfere with the original design of the SPK task.

File: espnet2/lid/loss/aamsoftmax.py (hunk 1 above)
Comment (sw005320) [reply espnet2/lid/loss/aamsoftmax.py]: It's more like a balance of 1) having each module separate (more easily modularized) or 2) avoiding duplicated code.
We can make a decision one by one.
For this particular function, I think it is OK for me to duplicate it with some comments about how they differ.

Comment (Jungjee) [reply espnet2/lid/loss/aamsoftmax.py]: I'm fine with however we proceed as long as @sw005320 is fine with it.
But, @Qingzheng-Wang , please note that the same losses in SPK side is also used for classification only because they are used only in the training phase. The way I see it, having acc printed out wouldn't harm SPK side.

More specifically,  

> This implementation returns predicted labels (via argmax) during inference, enabling output with LID predictions during inference.

Has nothing to do with SPK because in ESPnet-SPK, loss functions are detached in the first place

> It also computes and returns accuracy during training and evaluation, making it easier to monitor model performance.

Again, returning accuracy during training would be a plus for SPK side. 

Hope this clarifies!

Comment (sw005320) [reply espnet2/lid/loss/aamsoftmax.py]: @Qingzheng-Wang, I think @Jungjee's points make sense to me.
So, I think we should follow his suggestion.

Comment (Qingzheng-Wang) [reply espnet2/lid/loss/aamsoftmax.py]: Sure, I will make the modifications to the losses and poolings directly under the SPK module. Thanks for the clarifications and suggestions.

Comment (Jungjee) [reply espnet2/lid/loss/aamsoftmax.py]: Thanks, @sw005320 and @Qingzheng-Wang!

**Comment by sw005320** [comment]
@Jungjee, could you review it again since this PR now modifies the SID-related sources based on your recommendation.

**Comment by Jungjee** [comment]
Let me take a look tonight!

File: espnet2/spk/espnet_model.py [hunk 5]
```diff
@@ -99,12 +101,12 @@ def forward(
...
         # 1. extract low-level feats (e.g., mel-spectrogram or MFCC)
         # Will do nothing for raw waveform-based models (e.g., RawNets)
-        feats, _ = self.extract_feats(speech, None)
+        feats, feat_lengths = self.extract_feats(speech, speech_lengths)
 
         frame_level_feats = self.encode_frame(feats)
 
         # 2. aggregation into utterance-level
-        utt_level_feat = self.pooling(frame_level_feats, task_tokens)
+        utt_level_feat = self.pooling(frame_level_feats, feat_lengths=feat_lengths)
```
Comment (Jungjee) [inline espnet2/spk/espnet_model.py:109]: is the removal of `task_tokens` okay for the SPK side?

File: espnet2/spk/loss/aamsoftmax.py [hunk 6]
```diff
@@ -15,65 +14,100 @@
...
+        nout: int,
+        nclasses: int,
+        margin: float = 0.3,
+        scale: int = 15,
+        easy_margin: bool = False,
+        **kwargs,
     ):
         super().__init__(nout)
 
-        self.test_normalize = True
```
Comment (Jungjee) [inline espnet2/spk/loss/aamsoftmax.py:34]: Doesn't this break the SPK side at test phase? 
Please let me know if you checked this.

**Review by Jungjee — COMMENTED** [review]
Thank you! 
I believe you would have tested already with LID.
But may I ask if you've tested the SPK side as well so that this PR doesn't break the SPK side? I was expecting something on the `spk/espnet_model.py` or relevant files to have some updates.

File: espnet2/spk/espnet_model.py (hunk 5 above)
Comment (Qingzheng-Wang) [reply espnet2/spk/espnet_model.py:109]: Yes. None of the three pooling functions require task_tokens, so its removal will not affect the SPK side.

File: espnet2/spk/loss/aamsoftmax.py (hunk 6 above)
Comment (Qingzheng-Wang) [reply espnet2/spk/loss/aamsoftmax.py:34]: The test cases for spk have passed, and I could not find any usage of `test_normalize`, so I removed it.

File: espnet2/lid/espnet_model.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        else:
+            lang_embd = utt_level_feat
+
+        return lang_embd
+
+    def collect_feats(
+        self,
+        speech: torch.Tensor,
+        speech_lengths: torch.Tensor,
+        lid_labels: torch.Tensor = None,
```
[the remark this answers is not in the document]
Comment (Qingzheng-Wang) [reply espnet2/lid/espnet_model.py]: Fixed.

File: espnet2/spk/pooling/chn_attn_stat_pooling.py [hunk 8]
```diff
@@ -24,31 +27,75 @@ def __init__(self, input_size: int = 1536):
...
+    ) -> torch.Tensor:
+        r"""Forward pass of channel-attentive statistical pooling.
+
+        Args:
+            x: Input feature tensor of shape (batch_size, feature_dim, seq_len)
+            feat_lengths: Optional tensor of shape (batch_size,) containing
+                          the valid length of each sequence before padding
+
+        Returns:
+            x: Utterance-level embeddings of shape (batch_size, 2 x feature_dim)
```
[the remark this answers is not in the document]
Comment (Qingzheng-Wang) [reply espnet2/spk/pooling/chn_attn_stat_pooling.py]: Fixed.

**Comment by Qingzheng-Wang** [comment]
> Thank you! I believe you would have tested already with LID. But may I ask if you've tested the SPK side as well so that this PR doesn't break the SPK side? I was expecting something on the `spk/espnet_model.py` or relevant files to have some updates.

Thanks for your comments. 
Yes, I have tested it by training on VoxCeleb, and confirmed that it does not break the SPK side.

**Comment by sw005320** [comment]
Can you add a test for these new functions?

**Comment by Qingzheng-Wang** [comment]
> Can you add a test for these new functions?

Yes, I will add tests.

**Comment by sw005320** [comment]
OK
I just merged this PR since it took time, but please make sure to add the test for the new functions at your follow up PRs
